internal/cms/signed.go

```diff
@@ -0,0 +1,324 @@
+// information. The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed data.
+// If more than one signature is present, the successful validation of any signature
+// implies that the content in the parsed signed data is valid.
+// On successful verification, the list of signing certificates that successfully
+// verify is returned.
+// If all signatures fail to verify, the last error is returned.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) Verify(ctx context.Context, opts x509.VerifyOptions) ([][]*x509.Certificate, error) {
+	if len(d.SignerInfos) == 0 {
+		return nil, ErrSignerInfoNotFound
+	}
+	if len(d.Certificates) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	verifiedSignerMap := map[string][]*x509.Certificate{}
+	var lastErr error
+	for _, signerInfo := range d.SignerInfos {
+		signingCertificate := d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			lastErr = ErrCertificateNotFound
+			continue
+		}
+
+		certChain, err := d.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, signingCertificate.Raw)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = certChain
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertChains := make([][]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, certChain := range verifiedSignerMap {
+		verifiedSigningCertChains = append(verifiedSigningCertChains, certChain)
+	}
+	return verifiedSigningCertChains, nil
+}
+
+// VerifySigner verifies the signerInfo against the user specified signingCertificate.
+//
+// This function should be used when:
+//
+// 1. The certificates field of d is missing. This function allows the caller to provide
+// a signing certificate to verify the signerInfo.
+//
+// 2. The caller doesn't trust the signer identifier (unsigned field) of signerInfo
+// to identify signing certificate. This function allows such caller to use their trusted
+// signing certificate.
+//
+// Note: the intermediate certificates (if any) and root certificates in the verify
+// options MUST be set by the caller. The certificates field of d is not used in this function.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
```

Comment:
nit: This waning also applies to `Verify` function.

Reply:
Yes. The `Verify` function has the warning, too.